enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,295 @@
+environment variables, `INGRESS_SET_HTTP_HEADERS`, which specify the headers to set in the 
+global section of the haproxy.config.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpSetHeaders` field value.
+
+The values of `INGRESS_SET_HTTP_HEADERS` are comma-delimited lists of values of
+the form `name:value`.  (Colons and commas are not allowed in HTTP header
+names, so their use unambiguously delimits values in the environment variable.)
+The template translates these values into an appropriate `http-response set-header` stanzas.
+
+For example, if `ROUTER_SET_HTTP_HEADERS` is set to
+`X-Frame-Options: DENY,X-XSS-Protection: 1;mode=block`, then the template adds the following stanzas to
+the appropriate frontends:
+
+```haproxy.config's global section
+    http-response set-header X-Frame-Options: DENY
+    http-response set-header X-XSS-Protection: 1;mode=block    
+```
+
+This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
+[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).
+
+B) Setting HTTP headers in the backend section `default_backend openshift_default` of the haproxy.config via Route's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/router
+
+The router configures HAProxy using a configuration template. The template is populated with values in ServiceAliasConfig .
+The enhancement adds a field in ServiceAliasConfig called `RouteSetHTTPHeaders`, which specify the headers to set in the
+backend section `default_backend openshift_default` section of the haproxy.config.
+The router sets this variable based on the Route's
+`spec.httpSetHeaders` field value. Service AliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)
+
+The values of `RouteSetHTTPHeaders` are comma-delimited lists of values of
+the form `name:value`.  (Colons and commas are not allowed in HTTP header
+names, so their use unambiguously delimits values in the environment variable.)
+The template translates these values into an appropriate `http-response set-header` stanzas.
+
+For example, if `RouteSetHTTPHeaders` is set to
+`X-Frame-Options: DENY,X-XSS-Protection: 1;mode=block`, then the template adds the following stanzas to
+the appropriate frontends:
+
+```haproxy.config's backend section called default_backend openshift_default
+    http-response set-header X-Frame-Options: DENY
+    http-response set-header X-XSS-Protection: 1;mode=block    
+```
+
+[Code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L608) where this
+will be added.
```

Comment:
Use a permalink, or delete the link if it isn't needed.

Reply:
This has been changed.